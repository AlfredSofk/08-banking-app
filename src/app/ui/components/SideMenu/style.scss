.side-menu {
    position: fixed;
    top: 0;
    left: 0;
    width: 250px;
    height: 100vh;
    background-color: #2c3e50;
    color: #ecf0f1;
    padding-top: 1rem;
    display: flex;
    flex-direction: column;
    font-family: Arial, sans-serif;

    &__toggle {
        background: none;
        border: none;
        color: #ecf0f1;
        font-size: 1.5rem;
        cursor: pointer;
        padding: 0.5rem;
        margin-left: 1rem;

        &:hover {
            color: #3498db;
        }
    }

    &__list {
        list-style: none;
        padding: 0;
        margin: 0;
        flex-grow: 1;
    }

    &__item {
        margin-bottom: 1rem;
    }

    &__link {
        color: #ecf0f1;
        text-decoration: none;
        padding: 0.5rem 1rem;
        display: block;

        &:hover,
        &:focus {
            background-color: #34495e;
            color: #3498db;
        }
    }

    &__dropdown {
        list-style: none;
        padding: 0;
        margin: 0.5rem 0 0 1rem;
    }

    &__dropdown-item {
        margin-bottom: 0.5rem;
    }

    &__dropdown-link {
        color: #bdc3c7;
        text-decoration: none;
        padding: 0.3rem 1rem;
        display: block;
        font-size: 0.9rem;

        &:hover,
        &:focus {
            color: #ecf0f1;
        }
    }

    &__item--bottom {
        margin-top: auto;
        padding-bottom: 1rem;
    }
}