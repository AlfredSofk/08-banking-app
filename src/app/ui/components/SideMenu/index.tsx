import  { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { FaHome, FaUser, FaCog, FaQuestion, FaBars, FaChevronDown } from 'react-icons/fa';
import { CiBank } from 'react-icons/ci';
import './style.scss';
import { routerDef } from '../../../routes';
import { TransactionItem } from '../TransactionItem';

export const SideMenu = () => {
  const [isExpanded, setIsExpanded] = useState(true);
  const [isDropdownOpen, setIsDropdownOpen] = useState(false);

  const transactionContainer = routerDef[0].children
  const trasactionPaths = transactionContainer?.slice(1, -2)

  const navigate = useNavigate();

  const handleToggleSidebar = () => {
    setIsExpanded(!isExpanded);
  };

  const handleNavigate = (path: string) => {
    let newPath = `/home/${path}`;
    navigate(newPath);
  };

  return (
    <aside className={`sidebar ${isExpanded ? 'sidebar--expanded' : ''}`} role='navigation' aria-label="Barra lateral de navegación">
      <button className="sidebar__toggle" onClick={handleToggleSidebar} aria-label="Expandir menú">
        <FaBars />
      </button>
      <div className="sidebar__logo"><CiBank /><p>BankU</p></div>

      <ul className="sidebar__menu">
        <li className="sidebar__item">
          <button
            className="sidebar__button"
            aria-label="Inicio"
            onClick={() => handleNavigate('inicio')}
          >
            <FaHome />
            {isExpanded && <span className="sidebar__text">Inicio</span>}
          </button>
        </li>

        <li className="sidebar__item sidebar__item--dropdown">
          <button
            className="sidebar__button"
            onClick={() => setIsDropdownOpen(!isDropdownOpen)}
            aria-haspopup="true"
            aria-expanded={isDropdownOpen}
            data-testid="botonDropdown"
          >
            <FaUser />
            {isExpanded && <span className="sidebar__text">Transacciones</span>}
            {isExpanded && <FaChevronDown className="sidebar__icon--dropdown" />}
          </button>
          { isDropdownOpen && 
            isExpanded &&
            <ul className="sidebar__submenu" data-testid="submenu">
              {
                trasactionPaths?.map((route, index) => {
                  return (
                    <TransactionItem key={index} path={route.path} handleNavigate={handleNavigate} />
                  )
                })       
              }
            </ul>
          }
      
        </li>
        <li className="sidebar__item">
          <button
            className="sidebar__button"
            aria-label="Cuenta"
            onClick={() => handleNavigate('account')}
          >
            <FaCog />
            {isExpanded && <span className="sidebar__text">Cuenta</span>}
          </button>
        </li>

        <li className="sidebar__item">
          <button
            className="sidebar__button"
            aria-label="Ayuda"
            onClick={() => handleNavigate('about')}
          >
            <FaQuestion />
            {isExpanded && <span className="sidebar__text">Acerca de</span>}
          </button>
        </li>
      </ul>
    </aside>
  );
};
