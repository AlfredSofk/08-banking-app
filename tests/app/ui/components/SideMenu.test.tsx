import { fireEvent, render, screen,  } from '@testing-library/react';
import { SideMenu } from '../../../../src/app/ui/components/SideMenu/index';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { vi } from 'vitest';

// Mock de useNavigate
vi.mock('react-router-dom', async () => {
    const actual = await vi.importActual('react-router-dom');
    return {
        ...actual,
        useNavigate: vi.fn(),
    };
});

describe('test para el componente SideMenu', async () => {

    const {useNavigate} = await import('react-router-dom'); 

    let navigateMock: ReturnType<typeof vi.fn>;

    beforeEach(() => {
        navigateMock = vi.fn();
        (vi.mocked(useNavigate ) as unknown as ReturnType<typeof vi.fn>).mockReturnValue(navigateMock);
    });

    afterEach(() => {
        vi.clearAllMocks();
    });
  
    test('match snapshot', () => {
      const { asFragment } = render( 
            <MemoryRouter>
                <SideMenu />
            </MemoryRouter>
      );
      expect(asFragment()).toMatchSnapshot();
    });

    test('renderiza correctamente con el menú expandido', () => {
        render(
            <MemoryRouter>
                <SideMenu />
            </MemoryRouter>
        );

        // Verificar que el menú principal está presente
        expect(screen.getByRole('navigation', { name: /Barra lateral de navegación/i })).toBeInTheDocument();

        // Verificar que las opciones principales están renderizadas
        expect(screen.getByLabelText(/inicio/i)).toBeInTheDocument();
        expect(screen.getByLabelText(/cuenta/i)).toBeInTheDocument();
        expect(screen.getByLabelText(/ayuda/i)).toBeInTheDocument();
    });

    test('colapsa y expande el menú al hacer clic en el botón de toggle', () => {
        render(
            <MemoryRouter>
                <SideMenu />
            </MemoryRouter>
        );

        const toggleButton = screen.getByLabelText(/expandir menú/i);

        // Menú inicialmente expandido
        expect(screen.getByRole('navigation')).toHaveClass('sidebar--expanded');

        // Colapsar el menú
        fireEvent.click(toggleButton);
        expect(screen.getByRole('navigation')).not.toHaveClass('sidebar--expanded');

        // Expandir el menú nuevamente
        fireEvent.click(toggleButton);
        expect(screen.getByRole('navigation')).toHaveClass('sidebar--expanded');
    });

    test('navega correctamente al hacer clic en las opciones principales', () => {
        render(
            <MemoryRouter>
                <SideMenu />
            </MemoryRouter>
        );

        // Simular clic en "Inicio"
        fireEvent.click(screen.getByLabelText(/inicio/i));
        expect(navigateMock).toHaveBeenCalledWith('/home/inicio');

        // Simular clic en "Cuenta"
        fireEvent.click(screen.getByLabelText(/cuenta/i));
        expect(navigateMock).toHaveBeenCalledWith('/home/account');

        // Simular clic en "Ayuda"
        fireEvent.click(screen.getByLabelText(/ayuda/i));
        expect(navigateMock).toHaveBeenCalledWith('/home/about');
    });

    test('muestra y oculta las transacciones al hacer clic en el botón de dropdown', () => {
        render(
            <MemoryRouter>
                <SideMenu />
            </MemoryRouter>
        );

        const dropdownButton = screen.queryByTestId("botonDropdown")
        let subMenu = screen.queryByTestId("submenu") 
        
        // Verificar que las transacciones no están visibles inicialmente
        expect(subMenu).not.toBeInTheDocument();
        
        // Mostrar las transacciones
        fireEvent.click(dropdownButton);
        subMenu = screen.queryByTestId("submenu")
        expect(subMenu).toBeInTheDocument();

        const liTransaction =  subMenu.querySelectorAll('li')

        // Verificar transacciones específicas
        expect(liTransaction[0].querySelector('button')?.textContent).toContain('Retirar Cajero');
        expect(liTransaction[4].querySelector('button')?.textContent).toContain('Compra Web');


        // Ocultar las transacciones
        fireEvent.click(dropdownButton);
        expect(subMenu).not.toBeInTheDocument();
    });

    test('navega correctamente al seleccionar una transacción', () => {
        render(
            <MemoryRouter>
                <SideMenu />
            </MemoryRouter>
        );

        const dropdownButton = screen.queryByTestId("botonDropdown")

        // Mostrar las transacciones
        fireEvent.click(dropdownButton);
        const liTrasaction = screen.queryByTestId("submenu")?.querySelectorAll('li') 

        // Simular clic en una transacción
        const transactionItem = liTrasaction[0].querySelector('button') as HTMLButtonElement;
        fireEvent.click(transactionItem);

        expect(navigateMock).toHaveBeenCalledWith('/home/retiroATM');
    });
});