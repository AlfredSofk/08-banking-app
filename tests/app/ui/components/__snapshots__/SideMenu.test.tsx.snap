// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test para el componente SideMenu > match snapshot 1`] = `
<DocumentFragment>
  <aside
    aria-label="Barra lateral de navegación"
    class="sidebar sidebar--expanded"
    role="navigation"
  >
    <button
      aria-label="Expandir menú"
      class="sidebar__toggle"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
        />
      </svg>
    </button>
    <div
      class="sidebar__logo"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          id="Bank"
        >
          <path
            d="M19.505,17.943V10.362a1.491,1.491,0,0,0,1.39-1.12,1.468,1.468,0,0,0-.7-1.68l-7.45-4.3a1.521,1.521,0,0,0-1.49,0l-7.45,4.3a1.468,1.468,0,0,0-.7,1.68,1.487,1.487,0,0,0,1.45,1.12h.13v7.57h-.12a1.5,1.5,0,0,0,0,3h14.87a1.5,1.5,0,0,0,.07-2.989ZM4.555,9.362a.505.505,0,0,1-.25-.94l7.45-4.289a.474.474,0,0,1,.49,0L19.7,8.422a.5.5,0,0,1-.25.94Zm13.95,1v7.57H14.9v-7.57Zm-4.61,0v7.57h-3.61v-7.57Zm-4.61,0v7.57h-3.6v-7.57Zm10.15,9.57H4.565a.5.5,0,0,1-.5-.5.5.5,0,0,1,.5-.5h14.87a.5.5,0,0,1,.5.5A.5.5,0,0,1,19.435,19.932Z"
          />
        </g>
      </svg>
      <p>
        BankU
      </p>
    </div>
    <ul
      class="sidebar__menu"
    >
      <li
        class="sidebar__item"
      >
        <button
          aria-label="Inicio"
          class="sidebar__button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z"
            />
          </svg>
          <span
            class="sidebar__text"
          >
            Inicio
          </span>
        </button>
      </li>
      <li
        class="sidebar__item sidebar__item--dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="sidebar__button"
          data-testid="botonDropdown"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
            />
          </svg>
          <span
            class="sidebar__text"
          >
            Transacciones
          </span>
          <svg
            class="sidebar__icon--dropdown"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
            />
          </svg>
        </button>
      </li>
      <li
        class="sidebar__item"
      >
        <button
          aria-label="Cuenta"
          class="sidebar__button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
            />
          </svg>
          <span
            class="sidebar__text"
          >
            Cuenta
          </span>
        </button>
      </li>
      <li
        class="sidebar__item"
      >
        <button
          aria-label="Ayuda"
          class="sidebar__button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 384 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M202.021 0C122.202 0 70.503 32.703 29.914 91.026c-7.363 10.58-5.093 25.086 5.178 32.874l43.138 32.709c10.373 7.865 25.132 6.026 33.253-4.148 25.049-31.381 43.63-49.449 82.757-49.449 30.764 0 68.816 19.799 68.816 49.631 0 22.552-18.617 34.134-48.993 51.164-35.423 19.86-82.299 44.576-82.299 106.405V320c0 13.255 10.745 24 24 24h72.471c13.255 0 24-10.745 24-24v-5.773c0-42.86 125.268-44.645 125.268-160.627C377.504 66.256 286.902 0 202.021 0zM192 373.459c-38.196 0-69.271 31.075-69.271 69.271 0 38.195 31.075 69.27 69.271 69.27s69.271-31.075 69.271-69.271-31.075-69.27-69.271-69.27z"
            />
          </svg>
          <span
            class="sidebar__text"
          >
            Acerca de
          </span>
        </button>
      </li>
    </ul>
  </aside>
</DocumentFragment>
`;
